### PR TITLE
Move cublasLT workspace allocation to buffer assignment

### PR DESCRIPTION
**Pull request by shawnwang18**
Move cublasLT workspace allocation to buffer assignment to support command buffer lowering

### CODE REVIEW

**Review by cheshire — APPROVED** [review]

File: xla/stream_executor/cuda/cuda_blas_lt.cc [hunk 1]
```diff
@@ -401,23 +401,25 @@ absl::Status BlasLt::MatmulPlan::ValidateInputs(
 absl::Status BlasLt::MatmulPlan::DoMatmul(
     Stream* stream, const void* alpha, DeviceMemoryBase a, DeviceMemoryBase b,
     const void* beta, DeviceMemoryBase c, DeviceMemoryBase d,
-    const MatmulAlgorithm& algorithm, ScratchAllocator& scratch_allocator,
```
Comment (akuegel) [inline xla/stream_executor/cuda/cuda_blas_lt.cc:404]: Could you possible do the stream_executor changes in a backward-compatible way? Tensorflow is also using these methods, so better would be to create new methods that are used from XLA, then once that has been merged, update Tensorflow and remove the old methods.

**Review by akuegel — CHANGES_REQUESTED** [review]

File: xla/service/gpu/gemm_rewriter.cc [hunk 2]
```diff
@@ -2173,14 +2189,18 @@ class GemmWorkspaceRewriteVisitor : public DfsHloRewriteVisitor {
...
     }
-    workspace = std::min(workspace, operands_byte_size);
 
     // Append workspace buffer to instruction outputs.
-    std::vector<Shape> output_shapes = {instr->shape()};
+    std::vector<Shape> output_shapes = instr->shape().IsArray()
+                                           ? std::vector<Shape>{instr->shape()}
+                                           : instr->shape().tuple_shapes();
     output_shapes.emplace_back(ShapeUtil::MakeShape(S8, {workspace}));
     Shape output_shape = ShapeUtil::MakeTupleShape(output_shapes);
```
Comment (jprabhas) [inline xla/service/gpu/gemm_rewriter.cc:2205]: I dont see the output_shape being used. Am I missing something?

File: xla/service/gpu/gemm_algorithm_picker.cc [hunk 3]
```diff
@@ -149,6 +148,12 @@ class GemmAutotuner {
 
   absl::StatusOr<AutotuneResult> TuneGpuBlasLt(const HloInstruction* gemm,
                                                const GemmConfig& gemm_config) {
+    TF_ASSIGN_OR_RETURN(
+        auto workspace_shape,
+        ShapeUtil::TryGetSubshape(gemm->shape(),
+                                  {gemm->shape().tuple_shapes_size() - 1}));
+    TF_ASSIGN_OR_RETURN(auto workspace_buffer, CreateBuffer(*workspace_shape));
```
Comment (sgerrard) [inline xla/service/gpu/gemm_algorithm_picker.cc:155]: This seems to be creating an error: use of undeclared identifier 'CreateBuffer'; did you mean 'DataBuffer'?
  150 |     TF_ASSIGN_OR_RETURN(auto workspace_buffer, CreateBuffer(*workspace_shape));

File: xla/stream_executor/cuda/cuda_blas_lt.cc (hunk 1 above)
Comment (shawnwang18) [reply xla/stream_executor/cuda/cuda_blas_lt.cc:404]: I did the fix that keep both APIs of allocating through scratch allocator, or through pre-allocated workspace buffer.

File: xla/service/gpu/gemm_algorithm_picker.cc (hunk 3 above)
Comment (shawnwang18) [reply xla/service/gpu/gemm_algorithm_picker.cc]: Oh, the error is due to not rebasing to the latest code, should resolve now.

**Review by ezhulenev — APPROVED** [review]

**Review by akuegel — APPROVED** [review]

**Review by akuegel — APPROVED** [review]

**Review by akuegel — APPROVED** [review]

**Review by cheshire — APPROVED** [review]

**Review by akuegel — APPROVED** [review]

**Pull request closed without merging**